frontend/app.js

```diff
@@ -0,0 +1,510 @@
+      .then((data) => {
+        return data.json()
+      })
+      .then((json) => {
+        this.setState({
+          tasks: json.tasks
+        })
+      })
+  }
+
+  addTask(event) {
+    event.preventDefault();
+    var name = this.state.taskName
+    var description = this.state.taskDescription
+    if (name == '') {
+      alert('you must write something')
+    } else {
+
+      fetch('/task/', {
+        method: 'POST',
+        headers: {
+          'Accept': 'application/json',
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify({
+          name: name,
+          description: description
+        })
+      }).then((data) => {
+          return data.json()
+        })
+        .then((json) => {
+
+          this.setState({
+            childVisible: false,
+            tasks: json.tasks,
+            taskName: '',
+            taskDescription: '',
+            submitBar: false
+          })
+
+      })
+    }
+  }
+
+  removeTask(event) {
+    fetch('/task/', {
+      method: 'DELETE',
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        id: event
+      })
+    }).then((data) => {
+        return data.json()
+      })
+      .then((json) => {
+        this.setState({
+          tasks: json.tasks
+        })
+      })
+  }
+
+  completeTask(event) {
+    console.log(event)
+    fetch('/task/', {
+      method: 'PUT',
+      headers: {
+        'Accept': 'application/json',
+        'Content-Type': 'application/json',
+      },
+      body: JSON.stringify({
+        id: event
+      })
+    }).then((data) => {
+        return data.json()
+      })
+      .then((json) => {
+        this.setState({
+          tasks: json.tasks
+        })
+      })
+  }
+
+  onTaskChange(event) {
+    this.setState({
+      taskName: event.target.value,
+    });
+    if (event.target.value.length > 0) {
+      this.setState({
+        childVisible: true
+      })
+    } else {
+      this.setState({
+        submitBar: false,
+        childVisible: false,
+        taskDescription: ''
+      })
+    }
+  }
+
+  onDesciptionChange(event) {
+    this.setState({
+      taskDescription: event.target.value
+    });
+    if (event.target.value.length > 0) {
+      this.setState({
+        submitBar: true
+      })
+    } else {
+
+      this.setState({
+        submitBar: false,
+
+      })
+    }
+  }
+
+  changeTab(event) {
+    if (this.state.completedState === 1) {
+      this.setState({
+        completedState: 0
+      })
+    } else if (this.state.completedState === 0) {
+      this.setState({
+        completedState: 1
+      })
+    }
+  }
+
+    render() {
+      if (this.state.tasks && this.state.completedState === 0) {
+        return (
+          <div className='root-wrapper'>
+            <CompletedTab
+              tasks={this.state.tasks}
+              completeTask={this.completeTask.bind(this)}
```

Comment:
Calling `bind` in the render function is a big react no-no. React, as you're probably aware, does a diff on whatever the render function returns. The `bind` function returns a brand new function, which will cause react to do more work than it has to, which can apparently cause performance problems. Google `bind in render` if you're interested in reading about this in more detail.

There are two valid solutions this problem:

1. `bind` in the constructor.

```javascript
constructor(...args) {
  super(...args);
  this.completeTask = this.completeTask.bind(this)
}
```

2. Use the property initializer syntax when you define the function:

```
// instead of 
completeTask() {
  // ...
}

// do this
completeTask = () => {
  // ...
}
```